bigtable/client/internal/endian.cc

```diff
@@ -0,0 +1,53 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "bigtable/client/internal/endian.h"
+#include <assert.h>
+#include <vector>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+std::string NumericToBigEndian(std::uint64_t value) {
```

Comment:
This code is not portable, it assumes the platform is LittleEndian, which while true for many (or most) computers it is not true in general.

You need to detect if the platform is BigEndian:

https://stackoverflow.com/questions/8978935/detecting-endianness

and if so, just memcpy:

```C++
static_assert(std::numeric_limit<unsigned char>::digits == 8,
  "This library has not been ported to platforms where CHAR_BIT != 8");
std::string result(8);
#if PLATORM_IS_BIGENDIAN
std::memcpy(result.data(), &value, sizeof(value));
```
If the platform is LittleEndian you need to do a byte swap first, e.g.:

```
std::uint64_t swapped = __builtin_bswap64(value); // 
```

That `__builtin_bswap64(()` function is a GCC intrinsic, and those are not portable to other compilers.  Clang and MSVC have their own builtins.  


And then do the `memcpy()`.  Note that on platforms where there are no alignment requirements doing a `reinterpret_cast<>` might be faster than a `memcpy()`.

Reply:
You can use `constexpr` functions to detect endianness btw, and then specialize some templates to pick the right function, without having to use the pre-processor hack I outlined before.